src/sentry/tasks/check_monitors.py

```diff
@@ -2,25 +2,59 @@
 
 import logging
 
+from datetime import timedelta
 from django.utils import timezone
 
-from sentry.models import Monitor, MonitorStatus, MonitorType
+from sentry.models import CheckInStatus, Monitor, MonitorCheckIn, MonitorStatus, MonitorType
 from sentry.tasks.base import instrumented_task
 
 
 logger = logging.getLogger('sentry')
 
+TIMEOUT = timedelta(hours=12)
+
 
 @instrumented_task(name='sentry.tasks.check_monitors', time_limit=15, soft_time_limit=10)
-def check_monitors():
+def check_monitors(current_datetime=None):
+    if current_datetime is None:
+        current_datetime = timezone.now()
+
     qs = Monitor.objects.filter(
         type__in=[MonitorType.HEARTBEAT, MonitorType.CRON_JOB],
-        next_checkin__lt=timezone.now(),
+        next_checkin__lt=current_datetime,
     ).exclude(
         status=MonitorStatus.DISABLED,
-    )
+    )[:10000]
     for monitor in qs:
         logger.info('monitor.missed-checkin', extra={
             'monitor_id': monitor.id,
         })
         monitor.mark_failed()
+
+    # timeout any monitors still marked as in progress after X time
+    qs = MonitorCheckIn.objects.filter(
+        status=CheckInStatus.IN_PROGRESS,
+        date_updated__lt=current_datetime - TIMEOUT,
```

Comment:
Is the checkin timeout something that users will want to tweak?

Reply:
probably, but the intention is to not expose every possibility at this stage